client/src/redux/selectors.js

```diff
@@ -0,0 +1,220 @@
+export const partiallyCompletedChallengesSelector = state =>
+  userSelector(state).partiallyCompletedChallenges || [];
+export const completionCountSelector = state => state[MainApp].completionCount;
+export const currentChallengeIdSelector = state =>
+  state[MainApp].currentChallengeId;
+
+export const emailSelector = state => userSelector(state).email;
+export const isVariantASelector = state => {
+  const email = emailSelector(state);
+  // if the user is not signed in and the user info is not available.
+  // always return A the control variant
+  if (!email) return true;
+  return emailToABVariant(email).isVariantA;
+};
+export const isDonatingSelector = state => userSelector(state).isDonating;
+export const isOnlineSelector = state => state[MainApp].isOnline;
+export const isServerOnlineSelector = state => state[MainApp].isServerOnline;
+export const isSignedInSelector = state => !!state[MainApp].appUsername;
+export const isDonationModalOpenSelector = state =>
+  state[MainApp].showDonationModal;
+export const recentlyClaimedBlockSelector = state =>
+  state[MainApp].recentlyClaimedBlock;
+export const donationFormStateSelector = state =>
+  state[MainApp].donationFormState;
+export const signInLoadingSelector = state =>
+  userFetchStateSelector(state).pending;
+export const showCertSelector = state => state[MainApp].showCert;
+export const showCertFetchStateSelector = state =>
+  state[MainApp].showCertFetchState;
+export const shouldRequestDonationSelector = state => {
+  const completedChallenges = completedChallengesSelector(state);
+  const completionCount = completionCountSelector(state);
+  const canRequestProgressDonation = state[MainApp].canRequestProgressDonation;
+  const isDonating = isDonatingSelector(state);
+  const recentlyClaimedBlock = recentlyClaimedBlockSelector(state);
+
+  // don't request donation if already donating
+  if (isDonating) return false;
+
+  // a block has been completed
+  if (recentlyClaimedBlock) return true;
+
+  // a donation has already been requested
+  if (!canRequestProgressDonation) return false;
+
+  // donations only appear after the user has completed ten challenges (i.e.
+  // not before the 11th challenge has mounted)
+  if (completedChallenges.length < 10) {
+    return false;
+  }
+
+  // this will mean we have completed 3 or more challenges this browser session
+  // and enough challenges overall to not be new
+  return completionCount >= 3;
+};
+
+export const userTokenSelector = state => {
+  return userSelector(state).userToken;
+};
+
+export const showCodeAllySelector = state => {
+  return state[MainApp].showCodeAlly;
+};
+
+export const userByNameSelector = username => state => {
+  const { user } = state[MainApp];
+  // return initial state empty user empty object instead of empty
+  // object litteral to prevent components from re-rendering unnecessarily
+  // TODO: confirm if "initialState" can be moved here or action-types.js
+  return user[username] ?? initialState.user;
```

Comment:
I was overthinking it in the meeting.  Let's just use `{}` as the fallback.  If we get this migrated to TS, we can revisit it then.

Also, we already do this once in `userSelector` :man_shrugging: